--- kernel-untouched/linux-3.19.0/include/linux/kvm_host.h	2016-01-14 15:17:52.000000000 -0800
+++ kernel-tlbsplit/linux-3.19.0/include/linux/kvm_host.h	2015-12-29 18:52:43.953940405 -0800
@@ -363,6 +363,7 @@
 	struct mutex slots_lock;
 	struct mm_struct *mm; /* userspace tied to this vm */
 	struct kvm_memslots *memslots;
+	struct kvm_splitpages *splitpages;
 	struct srcu_struct srcu;
 	struct srcu_struct irq_srcu;
 #ifdef CONFIG_KVM_APIC_ARCHITECTURE
